False-positive for wildcard generics

This started when updating chrome from 0.12.14 -> 0.13.0.

```
import org.chromium.build.annotations.Nullable;
import org.chromium.build.annotations.NullMarked;

@NullMarked
class Foo<OuterT> {

  interface Supplier<T extends @Nullable Object> {
    T get();
  }

  Supplier<@Nullable OuterT> sup;

  <T extends Supplier<?>> T acceptSup(T supplier) {
    return supplier;
  }

  void test() {
    acceptSup(sup);
  }
}
```

```
Foo.java:18: warning: [NullAway] incompatible types: Supplier<@org.chromium.build.annotations.Nullable OuterT> cannot be converted to Supplier<OuterT>
    acceptSup(sup);
              ^
    (see http://t.uber.com/nullaway )
```

Since `acceptSup` takes a `T extends Supplier<?>`, I think T should accept @Nullable generics.

There is no warning if passing a `Supplier<@Nullable String>`, nor if `acceptSup` takes a `Supplier<?>` (doesn't use generics).

--- a/nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java
+++ b/nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java
@@ -1965,9 +1965,21 @@ public final class GenericsChecks {
   }
 
   /**
-   * For some calls, javac drops nested type-use nullability annotations in inferred substitutions
-   * for method type variables. Recover these annotations from the corresponding actual argument
-   * types, while preserving one consistent top-level substitution per method type variable.
+   * In narrow cases, javac drops nested type-use nullability annotations on type variables in its
+   * inferred type for a generic method at a call site. See
+   * https://github.com/uber/NullAway/issues/1455. This method aims to restore those annotations
+   * based on the types of actual parameters. It does not attempt to be a very general fix, as we do
+   * not fully understand the scenarios where this can arise.
+   *
+   * @param invocationTree the method invocation tree for the generic method call
+   * @param origMethodType the declared method type for the generic method (to identify formal
+   *     parameters whose type is a type variable of the method)
+   * @param methodTypeAtCallSite the method type for the generic method as inferred by javac at the
+   *     call site
+   * @param state the visitor state
+   * @return a method type based on {@code methodTypeAtCallSite} but with some nested nullability
+   *     annotations on type variables restored to match those on actual parameters passed at the
+   *     call site
    */
   @SuppressWarnings("ReferenceEquality")
   private Type.MethodType restoreNestedNullabilityForTypeVarArguments(
@@ -1977,66 +1989,66 @@ public final class GenericsChecks {
       VisitorState state) {
     Symbol.MethodSymbol methodSymbol = ASTHelpers.getSymbol(invocationTree);
     if (methodSymbol.isVarArgs()) {
-      // TODO handle varargs methods
+      // skip handling of varargs for now
       return methodTypeAtCallSite;
     }
-    com.sun.tools.javac.util.List<Type> origArgTypes = origMethodType.getParameterTypes();
-    com.sun.tools.javac.util.List<Type> callSiteArgTypes = methodTypeAtCallSite.getParameterTypes();
-    List<? extends ExpressionTree> callArgs = invocationTree.getArguments();
-    if (origArgTypes.size() != callSiteArgTypes.size() || callArgs.size() != origArgTypes.size()) {
-      return methodTypeAtCallSite;
-    }
+    com.sun.tools.javac.util.List<Type> genericMethodParamTypes =
+        origMethodType.getParameterTypes();
+    com.sun.tools.javac.util.List<Type> callSiteParamTypes =
+        methodTypeAtCallSite.getParameterTypes();
+    List<? extends ExpressionTree> actualParams = invocationTree.getArguments();
 
     // use this map to store repaired substitutions for method type variables, to ensure we use the
-    // same repaired
-    // substitution for all occurrences of the same method type variable
+    // same repaired substitution for all occurrences of the same type variable
     Map<Symbol.TypeVariableSymbol, Type> repairedTopLevelSubstitutions = new HashMap<>();
     ListBuffer<Type> updatedArgTypes = new ListBuffer<>();
     boolean changed = false;
-    for (int i = 0; i < origArgTypes.size(); i++) {
-      Type updatedType = callSiteArgTypes.get(i);
-      Type origArgType = origArgTypes.get(i);
-      if (origArgType instanceof Type.TypeVar typeVar
-          && typeVar.tsym.owner == methodSymbol
-          && !(updatedType instanceof Type.TypeVar)) {
+    for (int i = 0; i < genericMethodParamTypes.size(); i++) {
+      Type callSiteParamType = callSiteParamTypes.get(i);
+      Type genericMethodParamType = genericMethodParamTypes.get(i);
+      // only attempt a repair when the generic method's parameter type is a type variable of the
+      // method
+      if (genericMethodParamType instanceof Type.TypeVar typeVar
+          && typeVar.tsym.owner == methodSymbol) {
         Symbol.TypeVariableSymbol typeVarSymbol = (Symbol.TypeVariableSymbol) typeVar.tsym;
         Type repairedSubstitution = repairedTopLevelSubstitutions.get(typeVarSymbol);
         if (repairedSubstitution != null) {
-          if (!state
-              .getTypes()
-              .isSameType(
-                  state.getTypes().erasure(repairedSubstitution),
-                  state.getTypes().erasure(updatedType))) {
-            // Inconsistent substitution for the same top-level type variable; bail out.
-            return methodTypeAtCallSite;
-          }
-          if (repairedSubstitution != updatedType) {
+          // re-use the previous substitution, to ensure consistency
+          if (repairedSubstitution != callSiteParamType) {
             changed = true;
-            updatedType = repairedSubstitution;
+            callSiteParamType = repairedSubstitution;
           }
         } else { // need to compute the substitution
-          Type actualArgType = getTreeType(callArgs.get(i), state);
+          Type actualArgType = getTreeType(actualParams.get(i), state);
+          // only handle cases of non-raw actual parameter types that have the same base type as the
+          // inferred parameter type at the call site
           if (actualArgType != null
               && !actualArgType.isRaw()
               && state
                   .getTypes()
                   .isSameType(
                       state.getTypes().erasure(actualArgType),
-                      state.getTypes().erasure(updatedType))) {
+                      state.getTypes().erasure(callSiteParamType))) {
+            // restore explicit nested annotations from the actual parameter type to the call site
+            // parameter type (this will only apply to nested type variables within
+            // callSiteParamType)
             Type restoredType =
                 TypeSubstitutionUtils.restoreExplicitNullabilityAnnotations(
-                    actualArgType, updatedType, config, Collections.emptyMap());
+                    actualArgType, callSiteParamType, config, Collections.emptyMap());
+            // remember the substitution so we use it consistently at other parameter positions
             repairedTopLevelSubstitutions.put(typeVarSymbol, restoredType);
-            if (restoredType != updatedType) {
+            if (restoredType != callSiteParamType) {
               changed = true;
-              updatedType = restoredType;
+              callSiteParamType = restoredType;
             }
           } else {
-            repairedTopLevelSubstitutions.put(typeVarSymbol, updatedType);
+            // remember that we did _not_ change anything, again for consistency across parameter
+            // positions
+            repairedTopLevelSubstitutions.put(typeVarSymbol, callSiteParamType);
           }
         }
       }
-      updatedArgTypes.append(updatedType);
+      updatedArgTypes.append(callSiteParamType);
     }
     if (!changed) {
       return methodTypeAtCallSite;

--- a/nullaway/src/test/java/com/uber/nullaway/jspecify/GenericMethodTests.java
+++ b/nullaway/src/test/java/com/uber/nullaway/jspecify/GenericMethodTests.java
@@ -1551,6 +1551,16 @@ public class GenericMethodTests extends NullAwayTestsBase {
               void test() {
                 acceptSup(sup);
               }
+              <T extends Supplier<?>> void acceptTwoSup(T supplier1, T supplier2) {
+              }
+              Supplier<OuterT> sup2 = make2();
+              Supplier<OuterT> make2() {
+                throw new RuntimeException();
+              }
+              void test2() {
+                // BUG: Diagnostic contains: incompatible types: Supplier<OuterT> cannot be converted to @NonNull Supplier<@Nullable OuterT>
+                acceptTwoSup(sup, sup2);
+              }
             }
             """)
         .doTest();

